# Multiple Linear Regression with sklearn - Exercise Solution

You are given a real estate dataset. 

Real estate is one of those examples that every regression course goes through as it is extremely easy to understand and there is a (almost always) certain causal relationship to be found.

The data is located in the file: 'real_estate_price_size_year.csv'. 

You are expected to create a multiple linear regression (similar to the one in the lecture), using the new data. 

Apart from that, please:
-  Display the intercept and coefficient(s)
-  Find the R-squared and Adjusted R-squared
-  Compare the R-squared and the Adjusted R-squared
-  Compare the R-squared of this regression and the simple linear regression where only 'size' was used
-  Using the model make a prediction about an apartment with size 750 sq.ft. from 2009
-  Find the univariate (or multivariate if you wish - see the article) p-values of the two variables. What can you say about them?
-  Create a summary table with your findings

In this exercise, the dependent variable is 'price', while the independent variables are 'size' and 'year'.

Good luck!

## Import the relevant libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import f_regression

## Load the data

In [3]:
data = pd.read_csv('real_estate_price_size_year.csv')

In [4]:
data.head()

price     size  year
0  234314.144   643.09  2015
1  228581.528   656.22  2009
2  281626.336   487.29  2018
3  401255.608  1504.75  2015
4  458674.256  1275.46  2009

## Create the regression

### Declare the dependent and the independent variables

In [5]:
y = data['price']
x = data[['size','year']]

### Regression

In [6]:
reg = LinearRegression()
reg.fit(x,y)

LinearRegression()

### Find the intercept

In [7]:
intercept = reg.intercept_
reg.intercept_

-5772267.017463279

### Find the coefficients

In [8]:
coeff = reg.coef_
reg.coef_

array([ 227.70085401, 2916.78532684])

### Calculate the R-squared

In [9]:
reg.score(x,y)

0.7764803683276795

### Calculate the Adjusted R-squared

In [10]:
r2 = reg.score(x,y)
n = x.shape[0]
p = x.shape[1]

adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)


0.7718717161282502

In [ ]:
adjusted_r2

### Compare the R-squared and the Adjusted R-squared

0.7764803683276795 vs adjusted(0.7718717161282502)

### Compare the Adjusted R-squared with the R-squared of the simple linear regression

Answer...

### Making predictions

Find the predicted price of an apartment that has a size of 750 sq.ft. from 2009.

In [14]:
new_data = [[750,2009]]
reg.predict(new_data)

C:\Users\VINICIUS\anaconda3\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([258330.34465995])

### Calculate the univariate p-values of the variables

In [16]:
p_values = f_regression(x,y)[1]
p_values.round(3)

array([0.   , 0.357])

In [17]:
reg_summary = pd.DataFrame(data=x.columns.values,columns=['Features'])

In [18]:
reg_summary ['Coef'] = reg.coef_

In [19]:
reg_summary ['p-value'] = p_values.round(3)

### Create a summary table with your findings

In [20]:
reg_summary

Features         Coef  p-value
0     size   227.700854    0.000
1     year  2916.785327    0.357

Answer...